리액트 8기 양승민 5주차 과제 제출

rukeni opened the pull request:
## 반영사항
- 동작 흐름
  - red - green - refactor에 의거하여 TDD방식으로 코드 작성하였습니다.
  - red 단계에 들어서면 꼭 green 과 refactor단계까지 갈 수 있도록 노력하였습니다.
  - refactor가 이루어지면 테스트코드를 보강할지, 리팩토링을 좀 더 해볼지, 새로운 red를 찾을지 고민하였습니다.

- 작성 흐름
  - App.test.jsx에서부터 코딩 시작. 테스트코드를 짜고 기능을 구현하면서 관심사 별로 컴포넌트를 분리 하였습니다.
  - Happy Path위주로 우선 작성하고 테스트코드에 시간소요를 줄이기 위해 머리속에서 쉬운 방식으로 접근하려고 하였습니다.
 
## 현재 반영되지 않은사항
- redux 연동
- api 호출
- redux-thunk 연동
- 추가적인 케이스 테스트

## 왜 그렇게 변경했는지
- 지난번보다는 좀더 커밋을 TDD를 보여줄수 있도록 나누려고 노력했습니다.(그래도 잘 안됨..)

## 테스트
```npm run lint``` 통과
```npm run test``` 커버리지 테스트 100%, e2e 실패(api 미연동 상태)

rukeni committed:
Chore: @swc/jest 연동

rukeni committed:
Chore: eslint-plugin-simple-import-sort 설치

rukeni commented on package.json:
```diff
@@ -30,6 +32,7 @@
     "eslint-plugin-jsx-a11y": "^6.4.1",
     "eslint-plugin-react": "^7.23.2",
     "eslint-plugin-react-hooks": "^4.2.0",
+    "eslint-plugin-simple-import-sort": "^7.0.0",
```
import할때 공백을 주거나, Global 임포트와 Local 임포트 구분할때 자꾸 실수를 하여 문명의 과학(?)을 살짝 도입하였습니다..

https://github.com/lydell/eslint-plugin-simple-import-sort

hannut91 replied:
좋습니다 ㅎㅎ

hannut91 replied:
앗 근데, 단순히 알파벳 순서는 문맥을 전혀 고려하지 않기 때문에 저는 별로 선호하지 않는 것 같아요.

hannut91 replied:
앗 근데, 단순히 알파벳 순서는 문맥을 전혀 고려하지 않기 때문에 저는 별로 선호하지 않는 것 같아요.

rukeni replied:
> 앗 근데, 단순히 알파벳 순서는 문맥을 전혀 고려하지 않기 때문에 저는 별로 선호하지 않는 것 같아요.

두번 강조... 체크✅

rukeni committed:
Add: request 목록 추가

rukeni committed:
Test: 지역리스트가 나와야한다.

rukeni committed:
Add: 지역 리스트 추가

rukeni committed:
Refactor: useState에 연동

rukeni committed:
Test: dom에 존재하는지 보단 value로 id값을 버튼이 잘 가지고 있는지 본다

rukeni committed:
Test: fixture를 이용하여 짧게 테스트 코드 리팩토링

rukeni committed:
Chore: @testing-library/user-event설치

rukeni committed:
Test: region버튼을 클릭하면 V표시가 나와야 한다.

rukeni committed:
Add: 버튼을 클릭하면 V표시하는 핸들러 추가

rukeni committed:
Test: 지역 리스트 Regions로 분리

rukeni committed:
Refactor: Regions 파일로 분리

hannut91 commented on src/Regions.test.jsx:
```diff
@@ -0,0 +1,16 @@
+import Regions from './Regions';
+
+describe('Regions', () => {
+  it('renders regions list', () => {
```
`renders regions`만 해도 이해할 수 있을 것 같아요

rukeni replied:
> `renders regions`만 해도 이해할 수 있을 것 같아요

반영하였습니다.

rukeni committed:
Chore: eslint-plugin-simple-import-sort 제거

rukeni commented on package.json:
```diff
@@ -34,7 +34,6 @@
     "eslint-plugin-jsx-a11y": "^6.4.1",
     "eslint-plugin-react": "^7.23.2",
     "eslint-plugin-react-hooks": "^4.2.0",
-    "eslint-plugin-simple-import-sort": "^7.0.0",
```
바로 삭제😅

rukeni committed:
Test: 카테고리 리스트가 있어야한다.

rukeni committed:
Test: fixtures에 categories를 추가

rukeni committed:
Add: Categories컴포넌트 추가

rukeni committed:
Test: category를 누르면 V표시가 생긴다.

rukeni committed:
Add: Categories 컴포넌트 추가 및 핸들러 이름 구체적으로 수정

rukeni committed:
Modify: ul 태그를 컴포넌트 내부로 이동

rukeni committed:
Refactor: click handler 재사용할 수 있도록 리팩토링

rukeni commented:
`Categories`와 `Regions`의 구조가 같네요. 의존관계가 심화될것 같다는 생각이 들지만 한번, `Buttons`라는 이름으로 묶어서 재사용성 강화 시도해보겠습니다.

rukeni committed:
Test: 배열을 받아서 버튼 리스트를 만든다. 버튼은 눌리면 (V)표시가 생긴다.

rukeni committed:
Add: Buttons 컴포넌트 추가

rukeni committed:
Add: Buttons 컴포넌트 추가

rukeni committed:
Merge branch 'rukeni' of https://github.com/rukeni/react-week5-assignment-1 into rukeni

rukeni committed:
Test: Buttons 컴포넌트 줄바꿈

rukeni committed:
Refactor: Categories와 Regions 공통 컴포넌트 Buttons 적용

rukeni committed:
Test: 지역과 분류를 선택하면 레스토랑 목록이 나와야 한다.

rukeni committed:
Add: Restaurnats목록 추가

rukeni committed:
Test: 통과를 위해 findByText -> getByText로 수정

rukeni committed:
Text: region과 category를 선택하지 않으면 restaurant는 null이어야 한다.

rukeni committed:
Modify: id 초기값이 MIN_SAFE_INTEGER라서 null이 없으므로 id체크를 부등호로 수정

rukeni committed:
Test: Restaurants로 분리 Restaurants는 regionId와 categoryId가 있어야만 렌더링된다.

rukeni committed:
Test: fixtures에 restaurants추가

rukeni committed:
Add: App컴포넌트에서 분리하여 Restaurants컴포넌트 추가

rukeni commented:
현 시점에서 자연스럽게 커버리지가 100%가 되었네요. 그말은 HappyPath만 작성했다는 것인가..

rukeni committed:
Tmp: E2E 통과시키는 부산 객체 임시 추가

rukeni committed:
Test: 불필요한 'list'문구 제거

hannut91 commented:
TDD를 하면 커버리지 100%가 나올 수 밖에 없습니다 ㅎㅎ 잘하셨습니다

hannut91 commented on src/Restaurants.jsx:
```diff
@@ -0,0 +1,15 @@
+export default function Restaurants({ restaurants, currentRegionId, currentCategoryId }) {
+  return (
+    <>
+      {(currentCategoryId >= 0 && currentRegionId >= 0) && (
```
음 이게 어떤 의미인지 잘 이해하기 어렵네요. 둘 다 있는 경우라면
```js
      {(currentCategoryId && currentRegionId) && (
```

이렇게 작성해도 괜찮을 것 같아요

hannut91 commented on src/Restaurants.jsx:
```diff
@@ -0,0 +1,15 @@
+            <li key={restaurant.id}>
+              {restaurant.name}
+            </li>
+          ))}
```
구조 분해 할당을 사용해서 더 간단하게 표현해도 좋을 것 같군요

hannut91 reviewed (commented):
현재 익숙한 방식으로 먼저 문제를 해결해 주셨군요 ㅎㅎ 좋습니다. 

redux를 이용해서도 문제를 해결해 봅시다

hannut91 commented:
리덕스 버전을 새로운 풀리퀘스트로 보내셔도 됩니다 ㅎㅎ

rukeni commented:
> 리덕스 버전을 새로운 풀리퀘스트로 보내셔도 됩니다 ㅎㅎ

아 ㅋㅋ 그것도 좋은방법이네요. 어제 저를 찾는 사람이 많아서 진행을 못했습니다.

오늘 작업하여 push하겠습니다.

rukeni committed:
Add: 데이터를 담당할 store 추가

rukeni committed:
Test: reducer가 void로 호출되면 기존 state를 반환해야한다.

rukeni committed:
Add: reducer 기능 추가

rukeni committed:
Test: Action Type으로 정의되지 않은 Action은 initialState를 반환해야한다.

rukeni committed:
Test: fixtures 데이터 분리

fixture값이 장황해져서 분리하였습니다. 다음과 같이 분리하였습니다.

1. categories
2. regions
3. restaurants
4. initialState(초기 state)
5. state(값이 있다고 가정한 state)

rukeni committed:
Test: context 문구 수정

rukeni committed:
Rename: fixtures/initialState -> fixtures/emptyState

rukeni committed:
Test: setRegions는 새로운 state.regions를 만든다.

rukeni committed:
Add: action 타입 추가

1. setRegions: regions를 받아서 반영

2. setCategories: categories를 받아서 반영

rukeni committed:
Add: actionHandlers 추가

rukeni committed:
Test: setRegions 액션 함수 임포트

rukeni committed:
Test: action 함수 임포트

rukeni committed:
Test: react-redux mock 추가

1. useDispatch
2. useSelector

rukeni committed:
Add: id값 갱신하는 action 및 reducer 추가

dispatch를 통해 V표시를 해주기 위해서 regionId와 categoryId를 저장하는 액션 및 리듀서 함수를 작성했습니다.

1. setCurrentRegionId
2. setCurrentCategoryId

rukeni committed:
Refactor: 부모로부터 state가 아닌 regions와 currentRegionId만 가져오도록 수정

rukeni committed:
Add: Redux Provider 추가

rukeni committed:
Refactor: currentId값 string으로 변경

rukeni committed:
Refactor: useState제거하고 redux 연동

rukeni committed:
Test: id값 string으로 변경

rukeni committed:
Style: global 임포트와 local 임포트 구분 공백

rukeni committed:
Test: redux 구조로 테스트 코드 수정

rukeni committed:
Refactor: 관심사에 맞게 상태 위치 변경

1. App.jsx에서 상태코드 제거
2. 각 관심사에 맞게 핸들러와 데이터 위치 이동

rukeni commented:
문득 분리를 하고나니 `App.jsx`는 맡고 있는 역할이 부모역할 말고는 없고 그러다보니 `App.test.jsx`역시 어떤 주체가 되는 행동을 테스트하기 어려운 것 같습니다. 어떤 좋은 방법이 있을까요?

rukeni committed:
Test: redux적용된 테스트코드로 수정

rukeni committed:
Test: 각 관심사로 넘어간 테스트 로직 제거

rukeni committed:
Test: reducer 테스트 항목 추가

1. setCurrentRegionId
2. setCurrentCategoryId

rukeni committed:
Test: 버튼 테스트 변경: button의 value와 state.regions에 있는 id값이 일치하는지 체크

rukeni committed:
Test: 버튼 테스트 변경: button의 value와 state.regions에 있는 id값이 일치하는지 체크

rukeni committed:
Merge branch 'rukeni' of https://github.com/rukeni/react-week5-assignment-1 into rukeni

rukeni committed:
Test: 이벤트 핸들러 테스트 코드 리팩토링

1. Buttons에서는 각 handleClick이 호출될때 id값을 물고 있는지 체크

2. categories와 regions에서는 각 dispatch가 올바르게 액션함수를 호출하는지 체크

rukeni commented:
이제야 리덕스로 리팩토링하고 커버리지 100을 찍었습니다.

이제 thunk를 도입하면서 서버로 데이터 요청을 로직을 작성하겠습니다.

rukeni committed:
Add: redux-thunk 추가

rukeni committed:
Add: api 함수 생성

fetchRegions, fetchCategories, fetchREstaurants로 나누어 fetch()를 한번 래핑하였습니다.

rukeni committed:
Add: thunk로 작동시킬 함수action 타입 loadRegions 선언

rukeni committed:
Add: regions 비동기 처리 연동

rukeni committed:
Add: categories 호출 연동

1. fetchCategories()추가 -> 서버요청하는 fetch 래핑함수

2. loadCategories()추가 -> 비동기 요청을 상태에 반영하는 thunk액션 함수

3. categories에 있던 mock 데이터 빈배열로 수정

4. useEffect를 통해 데이터 로드

rukeni committed:
Add: Restaurants 비동기 호출 추가

1. setRestaurants 액션 타입 추가(이전에 작성을 안했음)

2. loadRestaurants() thunk 액션 함수 추가

3. state.restaurants 빈배열로 수정

4. useEffect로 비동기 호출 추가.
-> 의존성 배열에 currentRegionId와 currentCategoryId를 추가하여 둘 중 하나가 변경되면 이펙트함수가 실행되도록 작성하였습니다.

rukeni committed:
Fix: currentRegionId를 통해 regionName알아오는 로직 추가

region과 category가 둘다 id로 넘어가는줄 알고 착각하여 regionName이 아닌 id값을 넘겨서 오류 발생하여 name값 찾는 로직을 loadREstaurants에 넣었습니다.

rukeni committed:
Add: setRestaurants ActionHandler 추가

rukeni committed:
Fix: 무한 호출되는 버그 수정

rukeni committed:
Test: useDispatch() -> 직접 jest.fn()호출로 변경

rukeni committed:
Test: setRestaurants 리듀서 테스트 코드 추가

rukeni commented:
흠 api연동하고나니 비동기 호출했던 부분이 테스트가 필요하네요.
이건 또 어떻게 처리해야하나...

rukeni committed:
Move: 폴더 구조 개편

파일이 조금 많아져서 보기 편하게 폴더별로 분류 하였습니다.

각 관심사 별로 소스와 test파일을 묶었습니다. 그리고 나머지는 'app'이라는 폴더를 만들어서 컴포넌트와 관계없는 것들을 모았습니다.

hannut91 commented:
> 문득 분리를 하고나니 App.jsx는 맡고 있는 역할이 부모역할 말고는 없고 그러다보니 App.test.jsx역시 어떤 주체가 되는 행동을 테스트하기 어려운 것 같습니다. 어떤 좋은 방법이 있을까요?

사실 App.jsx는 하는 일이 없습니다. 하지만 연습용으로 테스트를 작성해 보세요. 강의 에서도 여기서 테스트를 작성하는 것이 거의 E2E테스트 가까이 되는데, 이럴거면 E2E테스트를 따로 짜는게 좋을거라고 말씀하시죠.

hannut91 commented on src/app/reducer.test.js:
```diff
@@ -0,0 +1,73 @@
+    });
+  });
+
+  describe('setCurrentCategoryId', () => {
+    it('set new currentCategoryId', () => {
+      const state = reducer(emptyState, setCurrentCategoryId('1'));
+
+      expect(state.currentCategoryId).toBe('1');
+    });
+  });
+
+  describe('setCurrentCategoryId', () => {
+    it('set new currentCategoryId', () => {
+      const state = reducer(emptyState, setCurrentCategoryId('1'));
+
+      expect(state.currentCategoryId).toBe('1');
+    });
+  });
```
여기 코드가 같은게 있네요

rukeni replied:
헛 반영하였습니다.

hannut91 commented on src/app/reducer.test.js:
```diff
@@ -0,0 +1,73 @@
+    });
+  });
+
+  describe('setRestaurants', () => {
+    it('set new restaurants into restaurants', () => {
+      const state = reducer(emptyState, setRestaurants(regions));
+
+      expect(state.restaurants).toEqual(regions);
+    });
+  });
```
`setRestaurants`인데 regions를 넣고 있네요

rukeni replied:
반영하였습니다.

hannut91 commented on src/Restaurants/Restaurants.jsx:
```diff
@@ -0,0 +1,33 @@
+    <>
+      {(currentCategoryId && currentRegionId) && (
+        <ul>
+          {restaurants.map((restaurant) => (
+            <li key={restaurant.id}>
+              {restaurant.name}
+            </li>
+          ))}
```
구조 분해 할당으로 간단하게 해 봅시다

rukeni replied:
적용하였습니다.

hannut91 commented:
`redux-mock-store`를 활용해서 테스트를 작성해 보세요.

See also
  - https://github.com/reduxjs/redux-mock-store

rukeni committed:
Test: 테스트 코드 수정

1. setCurrentCategoryId에대한 테스트 중복
-> setCurrentRegionId테스트 추가

2. setRestaurants를 테스트하는데 기대하는 객체가 regions였던 것
-> restaurants객체를 기대하도록 수정

rukeni committed:
Chore: redux-mock-store 설치

rukeni committed:
Refactor: 객체를 구조분해 할당으로 변경

rukeni committed:
Refactor: 비동기 액션함수 분리

rukeni committed:
Chore: msw 설치

rukeni committed:
Chore: msw 환경 셋팅

rukeni committed:
Fix: store import 위치 수정

rukeni committed:
Test: 비동기 액션 테스트 코드 추가

1. redux-mock-store 사용
2. fetch관련해서는 mock으로 대체

rukeni committed:
Test: 아이디 값으로 지역을 찾는다.

rukeni committed:
Add: findRegion 헬퍼 함수 추가

rukeni committed:
Refactor: 지역이름 가져오는 로직 헬퍼함수로 분리

rukeni commented:
## 반영사항
- 작성내용
  - redux-mock-store를 활용하여 store를 mocking하여 테스트 코드 작성하였습니다.
  - redux-thunk와 api호출을 도입하면서 생기는 추가적인 기능과 테스트코드를 작성하였습니다.
  - 찾아보면서 [msw](https://mswjs.io/) 라는 라이브러리를 알게되었는데 이해도가 부족하여 반영하지 못했습니다.

## 왜 그렇게 변경했는지
- 좀 더 의욕을 가지고 해볼려고 msw를 적극 설치하였으나, 아직 잘 사용하지 못하여 fetch함수를 mocking하여 redux-mock-store만 활용하여 작성하였습니다.

## 테스트
```npm run lint``` 통과
```npm run test``` 통과

hannut91 commented on src/app/asyncActions.js:
```diff
@@ -0,0 +1,30 @@
+  return async (dispatch, getState) => {
+    const { regions, currentRegionId, currentCategoryId } = getState();
+    // TODO: 아예 지역이름을 가져오는 대대적 수정이 필요..
+    const regionName = findRegion(regions, currentRegionId).name;
```
```js
const { name } = findRegion(regions, currentRegionId);
```

이렇게 해도 될 것 같긴한데, 찾을 수 없는 경우 곤란할 것 같네요. find는 항상 없는 경우를 고려해야됩니다.

hannut91 commented on src/Buttons/Buttons.test.jsx:
```diff
@@ -0,0 +1,27 @@
+import Buttons from './Buttons';
+
+describe('Buttons', () => {
+  it('renders Buttons', () => {
+    const handleClick = jest.fn();
+    const currentId = '1';
+
+    render(
+      <Buttons
+        items={state.regions}
+        targetName="currentRegionId"
+        currentId={currentId}
+        onClick={handleClick}
+      />,
+    );
+    const buttons = screen.getAllByRole('button');
+    buttons.forEach((button, index) => {
+      fireEvent.click(button);
+
+      expect(button).toHaveValue(String(state.regions[index].id));
+      expect(handleClick).toHaveBeenCalledWith(state.regions[index].id);
+    });
+  });
```
단순히 렌더만 테스트하는 것이 아니라 뭔가 많은 일을 하고 있는 것 같네요

rukeni replied:
흠... 뭔가 많은 일을 줄여보겠습니다.

hannut91 commented on src/Categories/Categories.test.jsx:
```diff
@@ -0,0 +1,36 @@
+    categories: state.categories,
+    currentCategoryId: state.currentCategoryId,
+  }));
+  it('renders Categories', () => {
+    const handleClick = jest.fn();
+
+    render(<Categories
+      categories={state.categories}
+      currentCategoryId={state.currentCategoryId}
+      onClick={handleClick}
+    />);
+
+    fireEvent.click(screen.getByText('한식'));
+    expect(dispatch).toBeCalledWith({
+      type: 'setCurrentCategoryId',
+      payload: {
+        currentCategoryId: 1,
+      },
+    });
+  });
```
이것도 단지 렌더를 테스트하기 보다는 무언가 하고 있는 것 같아요

rukeni replied:
이것도 테스트만 집중 하도록 수정해보겠습니다.

hannut91 commented on src/Regions/Regions.jsx:
```diff
@@ -0,0 +1,32 @@
+  useEffect(() => {
+    dispatch(loadRegions());
+  }, []);
+  return (
```
위에 빈 줄 하나를 추가해 주세요.

hannut91 commented on src/Regions/Regions.test.jsx:
```diff
@@ -0,0 +1,30 @@
+    regions: state.regions,
+    currentRegionId: state.currentRegionId,
+  }));
+  it('renders regions', () => {
```
여기도 위에 빈 줄 하나를 추가해 주세요

hannut91 commented on src/Regions/Regions.test.jsx:
```diff
@@ -0,0 +1,30 @@
+
+import { useDispatch, useSelector } from 'react-redux';
+
+import state from '../../fixtures/state';
+import Regions from './Regions';
```
컴포넌트 임포트와 fixture임포트를 구분해 주세요

hannut91 commented on src/Regions/Regions.test.jsx:
```diff
@@ -0,0 +1,30 @@
+    render(<Regions />);
+
+    fireEvent.click(screen.getByText('서울'));
+    expect(dispatch).toBeCalledWith({
```
여기도 위에 빈 줄 하나를 추가해서 단정문과 구분해 주세요

hannut91 commented on src/app/asyncActions.test.js:
```diff
@@ -0,0 +1,72 @@
+
+describe('loadCategories', () => {
+  it('execute fetch data then dispatch setCategories', async () => {
+    window.fetch = jest.fn();
+    window.fetch.mockReturnValue(Promise.resolve({
+      json: () => Promise.resolve([{ id: 1, name: '한식' }]),
+    }));
```
전역 객체를 모킹하는 것은 힘든 일이에요.  API호출 함수 자체를 모킹하는 것은 어떨까요

rukeni replied:
fetch를 mocking해야된다! 이생각에 사로잡혀서 fetch를 mocking하는 쪽으로만 고민한것 같습니다. 참고하여 수정해보겠습니다.

hannut91 commented:
과제풀이 영상을 통해 복습을 진행해 주세요. 특히 풀이 과정과 어떤 결정들을 내리는지 자세히 보세요

rukeni committed:
Test: 테스트 코드 간소화

1. Buttons에서 버튼 하나만 테스트하는 걸로 코드 수정

2. Categories dipsatch가 불렸는지만 확인하는걸로 수정

rukeni committed:
Styles: 공백추가하여 맥락 구분
